File: src/main/scala/com/lp/test/datastream/CustomParallelSourceFunction.scala
package com.lp.test.datastream

import org.apache.flink.streaming.api.functions.source.{ParallelSourceFunction, SourceFunction}

/**
  * <p/> 
  * <li>Description: 自定义source能够并行</li>
  * <li>@author: lipan@cechealth.cn</li> 
  * <li>Date: 2019-04-15 20:05</li> 
  */
class CustomParallelSourceFunction extends ParallelSourceFunction[Long] {

  var isRunning = true
  var count = 1L

  override def run(ctx: SourceFunction.SourceContext[Long]): Unit = {
    while (isRunning) {
      ctx.collect(count)
      count += 1
      Thread.sleep(1000)
    }
  }

  override def cancel(): Unit = {

    isRunning = false
  }
}

File: src/main/java/com/lp/java/demo/datastream/source/function/CustomParallelSourceFunction.java
package com.lp.java.demo.datastream.source.function;

import org.apache.flink.streaming.api.functions.source.ParallelSourceFunction;


/**
 * <p/>
 * <li>Description: 自定义并行流</li>
 * <li>@author: panli0226@sina.com</li>
 * <li>Date: 2019-04-15 13:28</li>
 * note: ParallelSourceFunction继承的SourceFunction接口，是怎样实现并行流的？
 * 在初始化DataStreamSource对象中，对是否是并行流进行了判断
 */
public class CustomParallelSourceFunction implements ParallelSourceFunction<Long> {
    private static final long serialVersionUID = -4479361562567758002L;
    boolean isRunning = true;
    long count = 1;

    @Override
    public void run(SourceContext<Long> ctx) throws Exception {
        while (true) {
            ctx.collect(count);
            count += 1;
            Thread.sleep(1000);
        }
    }

    @Override
    public void cancel() {
        isRunning = false;
    }
}
